pallets/handles/src/lib.rs

```diff
@@ -0,0 +1,628 @@
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+
+		/// AccountId truncated to 32 bytes
+		type ConvertIntoAccountId32: Convert<Self::AccountId, AccountId32>;
+
+		/// A type that will supply MSA related information
+		type MsaInfoProvider: MsaLookup + MsaValidator<AccountId = Self::AccountId>;
+
+		/// The minimum suffix value
+		#[pallet::constant]
+		type HandleSuffixMin: Get<u32>;
+
+		/// The maximum suffix value
+		#[pallet::constant]
+		type HandleSuffixMax: Get<u32>;
+
+		#[cfg(feature = "runtime-benchmarks")]
+		/// A set of helper functions for benchmarking.
+		type MsaBenchmarkHelper: MsaBenchmarkHelper<Self::AccountId>;
+	}
+
+	// Storage
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	/// - Keys: k1: Canonical base handle, k2: Suffix
+	/// - Value: MSA id
+	#[pallet::storage]
+	#[pallet::getter(fn get_msa_id_for_canonical_and_suffix)]
+	pub type CanonicalBaseHandleAndSuffixToMSAId<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		Handle,
+		Twox64Concat,
+		HandleSuffix,
+		MessageSourceId,
+		OptionQuery,
+	>;
+
+	/// - Key: MSA id
+	/// - Value: Display name
+	#[pallet::storage]
+	#[pallet::getter(fn get_display_name_for_msa_id)]
+	pub type MSAIdToDisplayName<T: Config> =
+		StorageMap<_, Twox64Concat, MessageSourceId, Handle, ValueQuery>;
+
+	/// - Value: Cursor u16
+	#[pallet::storage]
+	#[pallet::getter(fn get_current_suffix_index_for_canonical_handle)]
+	pub type CanonicalBaseHandleToSuffixIndex<T: Config> =
+		StorageMap<_, Blake2_128Concat, Handle, SequenceIndex, OptionQuery>;
+
+	#[derive(PartialEq, Eq)] // for testing
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Invalid handle encoding (should be UTF-8)
+		InvalidHandleEncoding,
+		/// Invalid handle byte length
+		InvalidHandleByteLength,
+		/// Invalid handle character length
+		InvalidHandleCharacterLength,
+		/// The handle name is reserved for chain use only
+		HandleIsNotAllowed,
+		/// The handle contains characters that are not allowed
+		HandleContainsBlockedCharacters,
+		/// The handle does not contain characters in the supported ranges of unicode characters
+		HandleDoesNotConsistOfSupportedCharacterSets,
+		/// Suffixes exhausted
+		SuffixesExhausted,
+		/// Invalid MSA
+		InvalidMessageSourceAccount,
+		/// Cryptographic signature failed verification
+		InvalidSignature,
+		/// The MSA already has a handle
+		MSAHandleAlreadyExists,
+		/// The MSA does not have a handle needed to retire it.
+		MSAHandleDoesNotExist,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Deposited when a handle is claimed. [MSA id, full handle in UTF-8 bytes]
+		HandleClaimed {
+			/// MSA id of handle owner
+			msa_id: MessageSourceId,
+			/// UTF-8 string in bytes
+			handle: Vec<u8>,
```

Comment:
Did you mean to use your alias Handle and your other event?

Reply:
Part of #1392